### ISSUE TITLE
'Connection closed' when open connection -> close connection -> open connection again

### ISSUE THREAD

**Issue by trompx** [phpredis/phpredis#635]
Hello,

I have 2 docker redis servers 'reads' only for reads and 'default' for writes & reads which is used for session and cache too and use https://github.com/vetruvet/laravel-phpredis/blob/master/src/Vetruvet/PhpRedis/Database.php for the connection.

Redis servers :
```
'default' => [
	'host'       => '192.168.33.101',
	'port'       => 6379,
	'timeout'    => 1,
],

'reads' => [
	'host'       => '192.168.33.101',
	'port'       => 6380,
	'timeout'    => 1,
],
```

In my app, when doing the following :
```
$redis = PhpRedis::connection('reads');
$test = $redis->get('test');
echo $test;
$redis->close();

$redis = PhpRedis::connection('default');
$redis->set('test', 'ok');
$redis->close();
```

Or without closing the first connection :
```
$redis = PhpRedis::connection('default');
$test = $redis->get('test');
echo $test;

$redis = PhpRedis::connection('default');
$test = $redis->set('test', 'ok');
echo $test;
$redis->close();
```

All works fine.

But When I try to open a new connection that has been opened and closed before like :
```
$redis = PhpRedis::connection('reads');
$test = $redis->get('test');
echo $test;
$redis->close();

$redis = PhpRedis::connection('default');
$redis->set('test', 'ok');
$redis->close();

$redis = PhpRedis::connection('reads');
$test = $redis->get('test');
echo $test;
$redis->close();
```

Or like :
```
$redis = PhpRedis::connection('default');
$redis->set('test', 'ok');
$redis->close();

$redis = PhpRedis::connection('default');
$redis->set('test', 'ok');
$redis->close();
```

Or like :
```
$redis = PhpRedis::connection('reads');
$test = $redis->get('test');
echo $test;
$redis->close();

$redis = PhpRedis::connection('reads');
$test = $redis->get('test');
echo $test;
$redis->close();
```

I get the error :
```
ErrorException in file.php line xxx: Connection closed
```

If a connection to a server has already been opened and closed in a php function, during the same request if trying to open it again in another function will result in the error too.

Besides, even if I $redis->close(), connections stack up with TIME_WAIT & CLOSE_WAIT :
```
netstat -na | grep 6379
tcp        0      0 172.17.0.89:58805       172.17.42.1:6379        TIME_WAIT
tcp        0      0 172.17.0.89:58718       172.17.42.1:6379        TIME_WAIT
tcp        1      0 172.17.0.89:57540       172.17.42.1:6379        CLOSE_WAIT
tcp        0      0 172.17.0.89:58581       172.17.42.1:6379        TIME_WAIT
tcp        1      0 172.17.0.89:57535       172.17.42.1:6379        CLOSE_WAIT
tcp        0      0 172.17.0.89:58907       172.17.42.1:6379        TIME_WAIT
```

Any idea how I could get rid of those CLOSE_WAIT and why I get the connection closed error ?

Thanks for your help

PS : sorry for opening and closing the issue yesterday but it was late, I was going crazy and could not replicate correctly the issue...

**Comment by trompx** [comment]
For now the only solution I found to get rid of this error was to not use $redis->close(). It looks like I don't have anymore CLOSE_WAIT connections, only TIME_WAIT, but it kind of feels wrong to not close the connection after executing all redis commands. Is there any other way to solve this issue ?

I saw another post about pconnect https://github.com/phpredis/phpredis/issues/328 so I am not sur if I should use this instead...

Any advice would be very welcome :)

Thanks

**Comment by michael-grunder** [comment]
Hey,

When I run the following script:

```php
<?php

$r = new Redis();
$r->connect('localhost', 6379);
$r->set('foo', 'bar1');
$r->close();

$r->connect('localhost', 6379);
$r->set('foo','bar2');

echo "foo = " . $r->get('foo') . "\n";
?>
```

I get the expected result:

```
➜  phpredis git:(develop) ✗ php ltests/openclose.php
foo = bar2
```

However, if I comment out the second ```connect``` logic, I can create the error:

```php
<?php

$r = new Redis();
$r->connect('localhost', 6379);
$r->set('foo', 'bar1');
$r->close();

//$r->connect('localhost', 6379);
$r->set('foo','bar2');

echo "foo = " . $r->get('foo') . "\n";
?>
```

```
➜  phpredis git:(develop) ✗ php ltests/openclose.php
PHP Fatal error:  Uncaught exception 'RedisException' with message 'Connection closed' in /Users/mgrunder/dev/phpredis/ltests/openclose.php:9
Stack trace:
#0 /Users/mgrunder/dev/phpredis/ltests/openclose.php(9): Redis->set('foo', 'bar2')
#1 {main}
  thrown in /Users/mgrunder/dev/phpredis/ltests/openclose.php on line 9
```

So the issue here, is likely that whatever docker is _actually_ doing when you run ```redis::connection('reads')``` in this context is not re-opening the connection.  I don't use docker so I'm not sure what code actually gets executed in this context but that would make sense.

Does that help at all?  :octocat:

Cheers
Mike

**Comment by trompx** [comment]
Hello @michael-grunder,

Sorry to get back this late but I struggled with this and other issues.

If I connect directly like you with 
```
$redis = new Redis();
$redis->connect(192.168.33.101, 6379, 0 or 1);
```
it works just fine so I don't think it is a docker problem, more something introduce by the laravel vetruvet package.

But I tried to find why it was failing without success. The createSingleClients function seems to do the exact same thing as $redis->connect(192.168.33.101, 6379, 0 or 1); so I don't know why it is failing.

For now I will use phpredis without close function and will let you know if I found a solution later.

Cheers,
Xavier
